### ISSUE TITLE
High Accuracy mode stops working when enabled via notification command and entering zome.home_expanded

### ISSUE THREAD

**Issue by madface303** [home-assistant/android#2102]
**Home Assistant Android version:**
2021.12.3-full

**Phone model:**
Samsung Galaxy S21+

**Description of problem:**
When i am outside zone.home_expanded and enable high accuracy mode via the notification command it stops its updates in the moment i enter the zone.home_expanded. The binary sensor is still on, the notification of the companion app is still visible but updates no more (or only after a few minutes). Turning the mode off and on in the app solves the problem. 
The geofence and high accuracy mode works really fine when only enabled/disabled by the app itself. For me it seems the app is struggeling when high accuracy mode is on and it tries to enable it in the moment i enter my geofence (lines at about 10:35:42 in the log).

[homeassistant_companion_log_0-2-2022_10-40-13.txt](https://github.com/home-assistant/android/files/7799549/homeassistant_companion_log_0-2-2022_10-40-13.txt)

**Comment by dshokouhi** [comment]
I believe for this fix we may need to set this variable: https://github.com/home-assistant/android/blob/afa092b3c01defa06d2b9bd87440450a98573ee7/app/src/full/java/io/homeassistant/companion/android/sensors/LocationSensorManager.kt#L108

appropriately when high accuracy mode is turned on/off

https://github.com/home-assistant/android/blob/afa092b3c01defa06d2b9bd87440450a98573ee7/app/src/full/java/io/homeassistant/companion/android/sensors/LocationSensorManager.kt#L255-L277

currently its only set when the initial setup is done which is what happens when high accuracy mode is triggered

**Comment by dshokouhi** [comment]
Sounds like you are able to easily reproduce this issue, if possible try to give this debug APK a test. It will install side by side the production version. If it still fails lets look at the logs again to see if anything changed. Make sure to install the `full` version.

https://github.com/home-assistant/android/actions/runs/1647660206

**Comment by majkers** [comment]
For me high accuracy mode stops working when enabled via notification command always and not only after leaving or entering any zone. It was working just fine in releases before 2021.12 and now it is not. I can see an update never or only after a long period of time even thougn it should update every 7 seconds (I have it in my settings) as I changed it from default 60 or whatever when I started to check what might be wrong..

**Comment by madface303** [comment]
I tried the debug-apk, but when onboarding it throws me a firebase error
and i cannot send or receive any notification with it. Is there another way
to enable high accuracy mode when outside the zone.home_expanded to test if
the set variable and this version will work?

Am Mo., 3. Jan. 2022 um 08:41 Uhr schrieb majkers ***@***.***

**Comment by majkers** [comment]
I can confirm that when high accuracy mode is turn on from sensor setting it works and it is not when turned on by a notification in prod apk version

**Comment by dshokouhi** [comment]
Sorry I totally forgot notifications don't work in debug builds so you guys won't be able to test the change. Based on the provided logs the fix should work as the variable was out of sync with what it should've been.

**Comment by dshokouhi** [comment]
Reopening as users have reported this is still not fixed and now the mode wont stay enabled when turned on via notification command.

https://community.home-assistant.io/t/problem-with-high-accuracy-mode-mode-freezes/373839/9?u=dshokouhi

Latest logs:

```
01-05 12:56:53.402 26728  5064 D MessagingService: From: 313041709873
01-05 12:56:53.403 26728  5064 D MessagingService: Message data payload: {title=turn_on, message=command_high_accuracy_mode}
01-05 12:56:53.403 26728  5064 D MessagingService: Processing device command
01-05 12:56:53.408 26728 26728 D HighAccLocService: Try starting high accuracy location service (Interval: 5s)...
01-05 12:56:53.409 26728 26728 D ForegrndServiceLauncher: Start service HighAccuracyLocationService
01-05 12:56:53.412 26728 26728 D ForegrndServiceLauncher: Service HighAccuracyLocationService was created. Start service
01-05 12:56:53.412 26728 26728 D HighAccLocService: High accuracy location service created -> onCreate
01-05 12:56:53.413 26728 26728 D HighAccLocService: High accuracy location service (Interval: 5s) started -> onStartCommand
01-05 12:56:53.433 26728 26995 D ConnectivityManager: StackLog: [android.net.ConnectivityManager.sendRequestForNetwork(ConnectivityManager.java:3925)] [android.net.ConnectivityManager.registerDefaultNetworkCallbackForUid(ConnectivityManager.java:4489)] [android.net.ConnectivityManager.registerDefaultNetworkCallback(ConnectivityManager.java:4456)] [android.net.ConnectivityManager.registerDefaultNetworkCallback(ConnectivityManager.java:4430)] [androidx.work.impl.constraints.trackers.NetworkStateTracker.startTracking(NetworkStateTracker.java:89)] [androidx.work.impl.constraints.trackers.ConstraintTracker.addListener(ConstraintTracker.java:71)] [androidx.work.impl.constraints.controllers.ConstraintController.replace(ConstraintController.java:99)] [androidx.work.impl.constraints.WorkConstraintsTracker.replace(WorkConstraintsTracker.java:106)] [androidx.work.impl.background.greedy.GreedyScheduler.schedule(GreedyScheduler.java:157)] [androidx.work.impl.Schedulers.schedule(Schedulers.java:119)] [androidx.work.impl.utils.EnqueueRunnable.scheduleWorkInBackground(EnqueueRunnable.java:135)] [androidx.work.impl.utils.EnqueueRunnable.run(EnqueueRunnable.java:95)] [androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)] [java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)] [java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)] [java.lang.Thread.run(Thread.java:920)]
01-05 12:56:53.444 26728 26886 D SensorWorker: Updating all Sensors.
01-05 12:56:53.445 26728 26803 I WM-Processor: Moving WorkSpec (36210e51-b6ac-46c2-9290-0bc46c3c4343) to the foreground
01-05 12:56:53.451 26728 26728 I WM-SystemFgDispatcher: Started foreground service Intent { act=ACTION_START_FOREGROUND cmp=io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService (has extras) }
01-05 12:56:53.458 26728 26886 D ActivitySM: Unregistering for activity updates.
01-05 12:56:53.458 26728 26886 D ActivitySM: Registering for activity updates.
01-05 12:56:53.559 26728 26739 I mpanion.androi: Background young concurrent copying GC freed 324836(16MB) AllocSpace objects, 15(428KB) LOS objects, 33% free, 34MB/52MB, paused 86us,43us total 122.660ms
01-05 12:56:53.559 26728 26739 W mpanion.androi: Reducing the number of considered missed Gc histogram windows from 151 to 100
01-05 12:56:53.592 26728 26728 D ForegrndServiceLauncher: Check if service HighAccuracyLocationService is running. Service running = true
01-05 12:56:53.594 26728 12457 D LocBroadcastReceiver: High accuracy mode disabled, because not in zone [zone.home_expanded]
01-05 12:56:53.599 26728 12457 D LocBroadcastReceiver: High accuracy mode parameters changed. Disable high accuracy mode.
01-05 12:56:53.607 26728 12457 D HighAccLocService: Try stopping high accuracy location service...
01-05 12:56:53.611 26728 12457 D ForegrndServiceLauncher: Stop service HighAccuracyLocationService
01-05 12:56:53.611 26728 26728 D ForegrndServiceLauncher: Service HighAccuracyLocationService was destroyed. Stop service
01-05 12:56:53.612 26728 12457 D LocBroadcastReceiver: Registering for location updates.
01-05 12:56:53.614 26728 26728 D HighAccLocService: High accuracy location service stopped -> onDestroy
01-05 12:56:53.618 26728 26728 I WM-SystemFgDispatcher: Stopping foreground service
01-05 12:56:53.624 26728 26728 D ConnectivityManager: StackLog: [android.net.ConnectivityManager.unregisterNetworkCallback(ConnectivityManager.java:4598)] [androidx.work.impl.constraints.trackers.NetworkStateTracker.stopTracking(NetworkStateTracker.java:112)] [androidx.work.impl.constraints.trackers.ConstraintTracker.removeListener(ConstraintTracker.java:86)] [androidx.work.impl.constraints.controllers.ConstraintController.replace(ConstraintController.java:97)] [androidx.work.impl.constraints.WorkConstraintsTracker.replace(WorkConstraintsTracker.java:106)] [androidx.work.impl.foreground.SystemForegroundDispatcher.onExecuted(SystemForegroundDispatcher.java:144)] [androidx.work.impl.Processor.onExecuted(Processor.java:304)] [androidx.work.impl.Processor$FutureListener.run(Processor.java:377)]
01-05 12:56:53.625 26728 26728 I WM-SystemFgDispatcher: Stopping foreground service
01-05 12:56:53.644 26728 26886 D UrlRepository: localUrl is: false and usesInternalSsid is: false
```

**Comment by dshokouhi** [comment]
@chriss158  any chance you can help us out here? I'm afraid of introducing yet another bug while attempting to fix this issue. I thought my last fix would do it but apparently not :( Theres still a bit of logic here that I am not understanding completely

**Comment by chriss158** [comment]
I will have a look at this 👍

**Comment by chriss158** [comment]
I am just wondering how the command_high_accuracy_mode command should work. 

I'm not sure if the command should force the high accuracy mode, no matter if I'm in a certain zone or connected to a certain bluetooth device or not.

At the moment the problem is that the zones are not up to date. I.e. as soon as a zone change is made, the high accuracy mode is turned off, because the app does not know in which zone he is currently located. 


Translated with www.DeepL.com/Translator (free version)

**Comment by dshokouhi** [comment]
So from what I have gathered from posts on Reddit and also github issues is that users were using this as a way to trigger high accuracy mode earlier than the automation would detect. Like the linked forum post when I reopened this issue that is exactly what the user was doing.

Here are the current scenarios I see:

1. if a user has no constraints then they are enabling/disabling using HA automations directly and wont have this issue
2. If a user has a constraint which is currently `false` but they force high accuracy mode on then we should maybe assume we are "in zone/connected" until geofence/connection says otherwise or the user forces it off. The problem here is that we perform the check on the constraints often so maybe its best that if a user does an override then we wait for the next override or until the actual constraint becomes `true` so that it begins to work as expected once it goes `false`?

**Comment by majkers** [comment]
But like I said before. Myself I have no settings about zones or BT devices but I high accuracy mode is nit working for me either when enabled by notification and works ok when enabled from setting. Please keep that in mind

**Comment by dshokouhi** [comment]
> But like I said before. Myself I have no settings about zones or BT devices but high accuracy mode is not working for me either when enabled by notification and works ok when enabled from setting. Please keep that in mind

can you look at the logs when you enable high accuracy mode via notification command please....make sure to keep high accuracy mode running for about 30 seconds to see if there is another issue there. I just tested this myself and I see updates being sent to HA every 5 seconds.

your issue might be entirely different as this issue is about having zones/bluetooth devices configured for the setting

**Comment by madface303** [comment]
I hope i can explain the logic in some simple words what i'm as a user expected the function to be:
When high_accuracy enabled per command it should stay as long as no other event happens (enter geofence, leave geofence, connect or disconnect bt for example). If one of the last happens it should do what normally happens, f.E. if turned on by command and entering geofence it stays on (as it would turn on if it wasn't enabled before). After an event happened it should follow it's normal logic like it is now.

**Comment by majkers** [comment]
OK I have logis from my 2 km drive. Update period set to 10 seconds and BT or Zones settings. Let me know if I should create new issue.
I turned on High accuracy mode with notification and there was no change of my location on HA companion map....

[homeassistant_companion_log_0-17-2022_8-4-12.txt](https://github.com/home-assistant/android/files/7879743/homeassistant_companion_log_0-17-2022_8-4-12.txt)

**Comment by dshokouhi** [comment]
we have had a few releases since this issue was created, does this issue still exist in the latest release that went out yesterday?

**Comment by madface303** [comment]
Just tried it, still the same. High accuracy starts and disables in the moment the first location is sent (when it recognizes that it is not in zone home expanded)
Heres the log:

```
03-15 19:39:00.726 26418 28689 D WebSocketRepository: Websocket: onMessage (text)
03-15 19:39:00.727 26418 28689 D WebSocketRepository: Message number 165 received: {"id":165,"type":"event","event":{"message":"command_high_accuracy_mode","title":"turn_on","data":{"ttl":0,"priority":"high"},"hass_confirm_id":"59b69848c92a7b22c5191f3039c43d80"}}
03-15 19:39:00.732 26418 26626 D WebSocketRepository: Sending message 235: {type=mobile_app/push_notification_confirm, webhook_id=9fc1847d0e67cc3baad438567ad1610da1d71b6xxxxxxxxxxc208562a185705d, confirm_id=59b69848cxxxxxxxxxx91f3039c43d80, id=235}
03-15 19:39:00.732 26418 26626 D WebSocketRepository: Message number 235 sent
03-15 19:39:00.771 26418 28689 D WebSocketRepository: Websocket: onMessage (text)
03-15 19:39:00.772 26418 28689 D WebSocketRepository: Message number 235 received: {"id":235,"type":"result","success":true,"result":null}
03-15 19:39:00.774 26418 28695 D MessagingService: Processing device command
03-15 19:39:00.778 26418 26418 D LocBroadcastReceiver: Removing background location requests.
03-15 19:39:00.780 26418 26418 D HighAccLocService: Try starting high accuracy location service (Interval: 5s)...
03-15 19:39:00.781 26418 26418 D ForegrndServiceLauncher: Start service HighAccuracyLocationService
03-15 19:39:00.783 26418 26418 D ForegrndServiceLauncher: Service HighAccuracyLocationService was created. Start service
03-15 19:39:00.783 26418 26418 D HighAccLocService: High accuracy location service created -> onCreate
03-15 19:39:00.784 26418 26418 D HighAccLocService: High accuracy location service (Interval: 5s) started -> onStartCommand
03-15 19:39:00.811 26418 26503 D ConnectivityManager: StackLog: [android.net.ConnectivityManager.sendRequestForNetwork(ConnectivityManager.java:3937)] [android.net.ConnectivityManager.registerDefaultNetworkCallbackForUid(ConnectivityManager.java:4501)] [android.net.ConnectivityManager.registerDefaultNetworkCallback(ConnectivityManager.java:4468)] [android.net.ConnectivityManager.registerDefaultNetworkCallback(ConnectivityManager.java:4442)] [androidx.work.impl.constraints.trackers.NetworkStateTracker.startTracking(NetworkStateTracker.java:89)] [androidx.work.impl.constraints.trackers.ConstraintTracker.addListener(ConstraintTracker.java:71)] [androidx.work.impl.constraints.controllers.ConstraintController.replace(ConstraintController.java:99)] [androidx.work.impl.constraints.WorkConstraintsTracker.replace(WorkConstraintsTracker.java:106)] [androidx.work.impl.background.greedy.GreedyScheduler.schedule(GreedyScheduler.java:157)] [androidx.work.impl.Schedulers.schedule(Schedulers.java:119)] [androidx.work.impl.utils.EnqueueRunnable.scheduleWorkInBackground(EnqueueRunnable.java:135)] [androidx.work.impl.utils.EnqueueRunnable.run(EnqueueRunnable.java:95)] [androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)] [java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)] [java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)] [java.lang.Thread.run(Thread.java:920)]
03-15 19:39:00.817 26418 28695 D SensorWorker: Updating all Sensors.
03-15 19:39:00.818 26418 26541 I WM-Processor: Moving WorkSpec (0b5abd61-60c3-4e28-8b41-499aca246583) to the foreground
03-15 19:39:00.823 26418 26418 D LocBroadcastReceiver: Received location update.
03-15 19:39:00.824 26418 26418 D LocBroadcastReceiver: Last Location: 
03-15 19:39:00.824 26418 26418 D LocBroadcastReceiver: Coords:(49.999999, 8.999999)
03-15 19:39:00.824 26418 26418 D LocBroadcastReceiver: Accuracy: 14.955
03-15 19:39:00.824 26418 26418 D LocBroadcastReceiver: Bearing: 0.0
03-15 19:39:00.825 26418 26418 D LocBroadcastReceiver: High accuracy mode disabled, because not in zone [zone.home_expanded]
```
